### ISSUE TITLE
How to block javascript injection in app webview.

### ISSUE THREAD

**Issue by lufengd3** [w3c/webappsec-csp#570]
[EvaluateJavascript](https://developer.android.com/reference/android/webkit/WebView#evaluateJavascript(java.lang.String,%20android.webkit.ValueCallback%3Cjava.lang.String%3E)) is a common way to implement jsbridge in mobile app dev, but it's also a injection risk for third party  webpage running in that  webview. I haven't found any CSP directive for this situation, I don't know if I have missed the directive or it  just  dosen't exist.

**Comment by jonathanKingston** [comment]
WebViews are acting as the browser on behalf of the user; they're outside of the CSP model for protection as are web extensions.

**Comment by lufengd3** [comment]
> WebViews are acting as the browser on behalf of the user; they're outside of the CSP model for protection as are web extensions.

Yes, the webview as same as browser in tec, but it's different in bussiness. For example, I made a shopping guide app, embed webview in my app with javascript injection, and then induced user login their amazon account to accomplish the order. Once user logined, I can do anything with my javascript. It's really easy to implement such a trick, but the web browser won't do this. Releated story by facebook: https://krausefx.com/blog/ios-privacy-instagram-and-facebook-can-track-anything-you-do-on-any-website-in-their-in-app-browser .

In this situation webpage content is not secure. You can not tell your user with a homepage banner: "Do not login in with any other app but chome". 

Except security problem, code injection also make website developer crazy: where the script error come from, I have never write those code from error stack.

In conclusion, I think the CSP model should consider webview, as well as web extension code injection.

**Comment by dveditz** [comment]
The problem you describe is real, and apps that do this kind of snooping are reprehensible. Malicious apps are a bane. When a company trusted by billions of people (with a 'b', literally) can do this it's especially galling. I hope folks in jurisdictions with good data privacy laws sue the pants off them. Honestly, my advice to people is to use your mobile browser as much as possible; don't use apps unless there's no alternative.

Unfortunately this is not a risk CSP was designed to defend against: it has to trust the application that is hosting it.

Your proposed feature would be implemented in the browser engine made by the companies who also make the WebView API. If they wanted, they could remove the EvaluateJavascript() API entirely. But they can't, really, because that is how legit applications interact with the web view content. They could, if they wanted, make their web views support some "don't inject me, bro!" signal separate from CSP. It could be something like the (ugly)`x-frame-options: deny` HTTP header, except instead of not loading the content it prevents injections. But if web content could use a feature like that or equivalent CSP to disable injected app scripts unilaterally it would break many non-malicious apps that users specifically chose and want to work—including all non-Safari browsers on iOS!!  

In a better world you might imagine only general-purpose browser apps would be allowed to inject scripts into arbitrary sites, and apps that were something else and happened to allow "browsing around" as a sub-functionality wouldn't be allowed to inject. But that would break lots of legit apps, too, that load pages from a website associated with the app (and only those pages) and use that injection to communicate between the web pages and the app (it may not even look like a "web page" when it's in the app). You'd also still have to deal with malicious apps claiming to be a "browser". 

There are a lot of competing interests—user, app-maker, web site author, app-store company—which makes for a very complex set of issues to solve. The smart folks on the mobile OS security teams will have to figure this out. This is not something we can balance unilaterally with CSP.

> In conclusion, I think the CSP model should consider [...] web extension code injection

It has, and the spec says user agents (browsers) should allow web extension injection because they are presumed to work on behalf of the user.  There are, of course, malicious extensions, but solving the problem of installing malicious code is outside the scope of CSP (the whole browser could be malicious, for that matter, or malware has installed a MITM proxy to inject content).

**Comment by lufengd3** [comment]
> 

Thank you for your detailed reply 👍
